File: app/src/main/java/com/pleon/buyt/ui/adapter/StoreListAdapter.java
package com.pleon.buyt.ui.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.pleon.buyt.R;
import com.pleon.buyt.model.Store;
import com.pleon.buyt.ui.BaseViewHolder;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView.Adapter;
import butterknife.BindView;
import butterknife.ButterKnife;

public class StoreListAdapter extends Adapter<StoreListAdapter.StoreHolder> {

    private Context context;
    private List<Store> stores;

    public StoreListAdapter(Context context) {
        this.context = context;
        // setHasStableIds is an optimization hint that you give to the RecyclerView
        // and tell it "when I provide a ViewHolder, its id is unique and will not change."
        setHasStableIds(true);
    }

    @NonNull
    @Override
    public StoreHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.store_list_row, parent, false);
        return new StoreHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull StoreHolder holder, int position) {
        // keep this method as lightweight as possible as it is called for every row
        if (stores != null) {
            Store store = stores.get(position);
            holder.categoryImgVi.setImageResource(store.getCategory().getStoreImageRes());
            holder.nameTxVi.setText(store.getName());
            holder.delRevealView.setAlpha(0f); // for the case of undo of deleted item
        } // else: case of data not being ready yet; set a placeholder or something
    }

    @Override
    public int getItemCount() {
        return stores == null ? 0 : stores.size();
    }

    // setHasStableIds() should also be set (in e.g. constructor). This is an optimization hint that you
    // give to the RecyclerView and tell it "when I provide a ViewHolder, its id is unique and won't change."
    @Override
    public long getItemId(int position) {
        return stores.get(position).getStoreId();
    }

    public void setStores(List<Store> stores) {
        this.stores = stores;
        notifyDataSetChanged();
    }

    public Store getStore(int position) {
        return stores.get(position);
    }

    // Adapter (and RecyclerView) work with ViewHolders instead of direct Views.
    class StoreHolder extends BaseViewHolder {

        @BindView(R.id.storeIcon) ImageView categoryImgVi;
        @BindView(R.id.storeName) TextView nameTxVi;

        StoreHolder(View itemView) {
            super(itemView); // the view (row layout) for the item
            ButterKnife.bind(this, itemView); // unbind() is required only for Fragments
        }
    }
}

File: app/src/main/kotlin/com/pleon/buyt/ui/adapter/StoreListAdapter.kt
package com.pleon.buyt.ui.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView.Adapter
import com.pleon.buyt.R
import com.pleon.buyt.database.dto.StoreDetail
import com.pleon.buyt.ui.BaseViewHolder
import kotlinx.android.synthetic.main.store_list_row.view.*

class StoreListAdapter(val cxt: Context) : Adapter<StoreListAdapter.StoreHolder>() {

    var stores = listOf<StoreDetail>()
        set(stores) {
            field = stores
            notifyDataSetChanged()
        }

    init {
        // setHasStableIds is an optimization hint that you give to the RecyclerView
        // and tell it "when I provide a ViewHolder, its id is unique and will not change."
        setHasStableIds(true)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): StoreHolder {
        val itemView = LayoutInflater.from(parent.context)
                .inflate(R.layout.store_list_row, parent, false)
        return StoreHolder(itemView)
    }

    override fun onBindViewHolder(holder: StoreHolder, position: Int) {
        // if (stores != null) {
        holder.bindStore(stores[position])
        // } else: case of data not being ready yet; set a placeholder or something
    }

    override fun getItemCount() = stores.size

    override fun getItemId(position: Int) = stores[position].store.storeId

    fun getStore(position: Int) = stores[position]

    // Adapter (and RecyclerView) work with ViewHolders instead of direct Views.
    inner class StoreHolder(view: View) : BaseViewHolder(view) {
        fun bindStore(storeDetail: StoreDetail) {
            itemView.storeIcon.setImageResource(storeDetail.store.category.storeImageRes)
            itemView.storeName.text = storeDetail.store.name
            itemView.purchaseCount.text = cxt.resources.getQuantityString(R.plurals.store_detail_purchase_count, storeDetail.purchaseCount, storeDetail.purchaseCount)
            itemView.totalSpending.text = cxt.resources.getQuantityString(R.plurals.store_detail_total_spending, storeDetail.totalSpending, storeDetail.totalSpending)
            itemView.circular_reveal.alpha = 0f // for the case of undo of deleted item
        }
    }
}
